File: zero-server/src/main/kotlin/com/zeroday/model/Skill.kt
package com.zeroday.model

import kotlinx.serialization.Serializable

/**
 * A single node in the skill tree. Skills are organised by [tree];
 * within a tree they form a directed acyclic graph declared in
 * [SkillService.trees]. Each skill has a [cost] in skill points and
 * optional [effects] applied via [com.zeroday.service.SkillService]
 * to the player's [Player] state.
 */
@Serializable
data class SkillDefinition(
    val id: String,
    val tree: SkillTree,
    val tier: Int,                 // 1..N; higher tier requires more prerequisite skills
    val name: String,
    val description: String,
    val cost: Int,                 // skill-point cost to unlock
    val prerequisites: List<String> = emptyList(),
    val effects: List<SkillEffect> = emptyList()
)

enum class SkillTree {
    OFFENSE,    // faster exploits, higher command crit
    DEFENSE,    // better firewall, lower detection
    INTEL,      // faster research, cheaper scripts
    INCOME      // more credits from tasks, ad rewards
}

/**
 * Effects are applied at the moment a skill is unlocked. They are
 * declarations only — the live multipliers live on [Player] and are
 * computed by [com.zeroday.service.SkillService.effectiveMultipliers].
 */
@Serializable
sealed class SkillEffect {
    abstract val kind: String

    @Serializable
    data class MultiplyCreditsEarned(val factor: Double) : SkillEffect() {
        override val kind: String = "multiply_credits_earned"
    }

    @Serializable
    data class MultiplyXpEarned(val factor: Double) : SkillEffect() {
        override val kind: String = "multiply_xp_earned"
    }

    @Serializable
    data class AddMaxResource(val cpu: Int = 0, val ram: Int = 0, val bandwidth: Int = 0) : SkillEffect() {
        override val kind: String = "add_max_resource"
    }

    @Serializable
    data class ReduceResourceCost(val factor: Double) : SkillEffect() {
        override val kind: String = "reduce_resource_cost"
    }

    @Serializable
    data class IncreaseRegenRate(val factor: Double) : SkillEffect() {
        override val kind: String = "increase_regen_rate"
    }
}

/**
 * The aggregated multipliers the player gets from their unlocked skills.
 * Cached per-player and invalidated on unlock.
 */
@Serializable
data class SkillMultipliers(
    val creditsEarned: Double = 1.0,
    val xpEarned: Double = 1.0,
    val resourceCost: Double = 1.0,
    val regenRate: Double = 1.0,
    val bonusCpu: Int = 0,
    val bonusRam: Int = 0,
    val bonusBandwidth: Int = 0
) {
    fun combine(other: SkillMultipliers) = SkillMultipliers(
        creditsEarned = creditsEarned * other.creditsEarned,
        xpEarned = xpEarned * other.xpEarned,
        resourceCost = resourceCost * other.resourceCost,
        regenRate = regenRate * other.regenRate,
        bonusCpu = bonusCpu + other.bonusCpu,
        bonusRam = bonusRam + other.bonusRam,
        bonusBandwidth = bonusBandwidth + other.bonusBandwidth
    )
}

File: zero-server-csharp/src/ZeroDayMMO.Shared/Models/Skill.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ZeroDayMMO.Shared.Models;

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum SkillTree
{
    [JsonPropertyName("offense")]
    OFFENSE,
    [JsonPropertyName("defense")]
    DEFENSE,
    [JsonPropertyName("intel")]
    INTEL,
    [JsonPropertyName("income")]
    INCOME
}

public class SkillDefinition
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = "";

    [JsonPropertyName("tree")]
    public SkillTree Tree { get; set; }

    [JsonPropertyName("tier")]
    public int Tier { get; set; }

    [JsonPropertyName("name")]
    public string Name { get; set; } = "";

    [JsonPropertyName("description")]
    public string Description { get; set; } = "";

    [JsonPropertyName("cost")]
    public int Cost { get; set; }

    [JsonPropertyName("prerequisites")]
    public List<string> Prerequisites { get; set; } = new();

    [JsonPropertyName("effects")]
    public List<SkillEffect> Effects { get; set; } = new();
}

[JsonPolymorphic(TypeDiscriminatorPropertyName = "kind")]
[JsonDerivedType(typeof(MultiplyCreditsEarned), typeDiscriminator: "multiply_credits_earned")]
[JsonDerivedType(typeof(MultiplyXpEarned), typeDiscriminator: "multiply_xp_earned")]
[JsonDerivedType(typeof(AddMaxResource), typeDiscriminator: "add_max_resource")]
[JsonDerivedType(typeof(ReduceResourceCost), typeDiscriminator: "reduce_resource_cost")]
[JsonDerivedType(typeof(IncreaseRegenRate), typeDiscriminator: "increase_regen_rate")]
public abstract class SkillEffect
{
    [JsonPropertyName("kind")]
    public abstract string Kind { get; }
}

public class MultiplyCreditsEarned : SkillEffect
{
    [JsonPropertyName("factor")]
    public double Factor { get; set; }

    [JsonPropertyName("kind")]
    public override string Kind => "multiply_credits_earned";
}

public class MultiplyXpEarned : SkillEffect
{
    [JsonPropertyName("factor")]
    public double Factor { get; set; }

    [JsonPropertyName("kind")]
    public override string Kind => "multiply_xp_earned";
}

public class AddMaxResource : SkillEffect
{
    [JsonPropertyName("cpu")]
    public int Cpu { get; set; } = 0;

    [JsonPropertyName("ram")]
    public int Ram { get; set; } = 0;

    [JsonPropertyName("bandwidth")]
    public int Bandwidth { get; set; } = 0;

    [JsonPropertyName("kind")]
    public override string Kind => "add_max_resource";
}

public class ReduceResourceCost : SkillEffect
{
    [JsonPropertyName("factor")]
    public double Factor { get; set; }

    [JsonPropertyName("kind")]
    public override string Kind => "reduce_resource_cost";
}

public class IncreaseRegenRate : SkillEffect
{
    [JsonPropertyName("factor")]
    public double Factor { get; set; }

    [JsonPropertyName("kind")]
    public override string Kind => "increase_regen_rate";
}

public class SkillMultipliers
{
    [JsonPropertyName("credits_earned")]
    public double CreditsEarned { get; set; } = 1.0;

    [JsonPropertyName("xp_earned")]
    public double XpEarned { get; set; } = 1.0;

    [JsonPropertyName("resource_cost")]
    public double ResourceCost { get; set; } = 1.0;

    [JsonPropertyName("regen_rate")]
    public double RegenRate { get; set; } = 1.0;

    [JsonPropertyName("bonus_cpu")]
    public int BonusCpu { get; set; } = 0;

    [JsonPropertyName("bonus_ram")]
    public int BonusRam { get; set; } = 0;

    [JsonPropertyName("bonus_bandwidth")]
    public int BonusBandwidth { get; set; } = 0;

    public SkillMultipliers Combine(SkillMultipliers other) => new()
    {
        CreditsEarned = CreditsEarned * other.CreditsEarned,
        XpEarned = XpEarned * other.XpEarned,
        ResourceCost = ResourceCost * other.ResourceCost,
        RegenRate = RegenRate * other.RegenRate,
        BonusCpu = BonusCpu + other.BonusCpu,
        BonusRam = BonusRam + other.BonusRam,
        BonusBandwidth = BonusBandwidth + other.BonusBandwidth
    };
}
